### PR TITLE
Feature/pw 403

### PR DESCRIPTION
**Pull request by Nikita-Polyakov**
https://soramitsu.atlassian.net/browse/PW-403
Add QR code for receive tokens functionality

### CODE REVIEW

File: src/components/AssetList.vue [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    AssetListItem,
+  },
+})
+export default class AssetList extends Mixins(TranslationMixin) {
+  @Prop({ default: () => [], type: Array }) readonly assets!: Array<Asset>;
+  @Prop({ default: 5, type: Number }) readonly size!: number;
+  @Prop({ default: false, type: Boolean }) readonly divider!: boolean;
+  @Prop({ default: false, type: Boolean }) readonly withFiat!: boolean;
+
+  wrapListeners(asset: Asset) {
```
Comment (stefashkaa) [inline src/components/AssetList.vue:54]: add a type for it please

File: src/components/QrCode.vue [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  @Prop({ default: '', type: String }) readonly value!: string;
+  @Prop({ default: 260, type: Number }) readonly size!: number;
+  @Ref('container') readonly container!: HTMLDivElement;
+
+  @Watch('value')
+  private rerender() {
+    this.renderCode();
+  }
+
+  element: SVGSVGElement | null = null;
```
Comment (stefashkaa) [inline src/components/QrCode.vue:22]: `Nullable<SVGSVGElement>`

File: src/components/QrCodeScanButton.vue [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+@Component
+export default class QrCodeScanButton extends Mixins(TranslationMixin) {
+  @Ref('input') readonly input!: HTMLInputElement;
+
+  openFileInput() {
+    this.input.click();
+  }
+
+  resetFileInput() {
```
Comment (stefashkaa) [inline src/components/QrCodeScanButton.vue:32]: `private` ?

File: src/components/QrCodeScanButton.vue [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+import TranslationMixin from './mixins/TranslationMixin';
+
+const reader = new BrowserQRCodeReader();
+
+@Component
+export default class QrCodeScanButton extends Mixins(TranslationMixin) {
+  @Ref('input') readonly input!: HTMLInputElement;
+
+  openFileInput() {
```
Comment (stefashkaa) [inline src/components/QrCodeScanButton.vue:28]: `: void`

File: src/components/QrCodeScanButton.vue (hunk 3 above)
Comment (stefashkaa) [reply src/components/QrCodeScanButton.vue:32]: `: void`

File: src/components/AssetList.vue [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      :style="style"
+    >
+      <template #before>
+        <div v-if="isEmptyList" class="asset-list-empty">
+          <slot name="list-empty">{{ t('assets.empty') }}</slot>
+        </div>
+      </template>
+
+      <template #default="{ item, index }">
+        <asset-list-item :asset="item" :withFiat="withFiat" :key="index" v-on="wrapListeners(item)">
```
Comment (stefashkaa) [inline src/components/AssetList.vue:18]: `:withFiat` -> `:with-fiat`

File: src/components/WalletSend.vue [hunk 6]
```diff
@@ -161,10 +163,43 @@ export default class WalletSend extends Mixins(
 
   step = 1;
   address = '';
+  assetBalance: Nullable<AccountBalance> = null;
+  assetBalanceSubscription: Nullable<Subscription> = null;
```
Comment (stefashkaa) [inline src/components/WalletSend.vue:167]: the same comment as above

File: src/components/WalletSend.vue [hunk 7]
```diff
@@ -348,6 +381,12 @@ export default class WalletSend extends Mixins(
     this.showAdditionalInfo = true;
     this.step = 3;
   }
+
+  resetAssetBalanceSubscription(): void {
```
Comment (stefashkaa) [inline src/components/WalletSend.vue:385]: `private`

File: src/lang/en/index.ts [hunk 8]
```diff
@@ -154,6 +154,14 @@ export default {
   },
   asset: {
     remove: 'Remove asset',
+    select: 'Select an asset',
```
Comment (stefashkaa) [inline src/lang/en/index.ts:157]: please add it to Lokalise when you'll add it to Polkaswap

File: src/util/image.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+export const IMAGE_MIME_TYPES = {
+  [IMAGE_EXTENSIONS.SVG]: 'image/svg+xml',
+  [IMAGE_EXTENSIONS.PNG]: 'image/png',
+  [IMAGE_EXTENSIONS.JPEG]: 'image/jpeg',
+};
+
+export const dataURItoBlob = (dataURI) => {
+  // convert base64 to raw binary data held in a string
```
Comment (stefashkaa) [inline src/util/image.ts:16]: do we need try-catch here?

File: src/util/image.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  JPEG = '.jpeg',
+}
+
+export const IMAGE_MIME_TYPES = {
+  [IMAGE_EXTENSIONS.SVG]: 'image/svg+xml',
+  [IMAGE_EXTENSIONS.PNG]: 'image/png',
+  [IMAGE_EXTENSIONS.JPEG]: 'image/jpeg',
+};
+
+export const dataURItoBlob = (dataURI) => {
```
Comment (stefashkaa) [inline src/util/image.ts:15]: please specify a type for `dataURI`

File: src/util/image.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+      } else {
+        reject(new Error(''));
+      }
+    };
+
+    image.onerror = (error) => reject(error);
+  });
+};
+
+export const createSvgBlob = (svgElement: SVGSVGElement) => {
```
Comment (stefashkaa) [inline src/util/image.ts:68]: please add return type here

**Comment by stefashkaa** [comment]
@Nikita-Polyakov do we have a color of QR code for light theme?
image

**Comment by stefashkaa** [comment]
@Nikita-Polyakov should it scan each asset OR only selected?

image

**Comment by stefashkaa** [comment]
@Nikita-Polyakov if user 1) doesn't have a token in account assets but he 2) has some amount and 3) he scanned that QR -> should we add this token into account assets after the transfer operation?

File: src/util/image.ts (hunk 9 above)
Comment (Nikita-Polyakov) [reply src/util/image.ts:16]: I think no, in worst case the error is throws and i catch it in component & show error notification

**Comment by Nikita-Polyakov** [comment]
> @Nikita-Polyakov do we have a color of QR code for light theme? Seems that it should be the same with title color image

Artem and I also explored different options, and came to the conclusion that the code in a color scheme other than black is not readable by scanners.

You can check such code with your mobile phone. We couldn't read it using our mobiles
qrcode-download-white

**Comment by Nikita-Polyakov** [comment]
> @Nikita-Polyakov if user 1) doesn't have a token in account assets but he 2) has some amount and 3) he scanned that QR -> should we add this token into account assets after the transfer operation?

I'm not sure, because if user has some amount of this asset, and it is not his assets list, may be he removed it specially

**Comment by Nikita-Polyakov** [comment]
> @Nikita-Polyakov do we have a color of QR code for light theme? Seems that it should be the same with title color image

**Comment by Nikita-Polyakov** [comment]
> @Nikita-Polyakov should it scan each asset OR only selected?
> 
> image

Now it processes all assets, this logic is transferred from the mobile application, there are no restrictions if you scan the code on the asset details screen

**Review by stefashkaa — APPROVED** [review]

**Pull request merged**